security/pkg/nodeagent/test/cert_rotation/rotation_test.go

```diff
@@ -21,63 +21,100 @@ import (
 	sdsTest "istio.io/istio/security/pkg/nodeagent/test"
+	"os/exec"
+)
+
+const (
+	rotateInterval = 1 * time.Second
+	proxyRunningTime = 4 * rotateInterval
+	sleepTime = 100 * time.Millisecond
+	httpSuccessCode = 200
+	certShouldUpdateCount = 2
+)
+
+var (
+	certSet = make(map[string]bool)
 )
 
 func TestCertRotation(t *testing.T) {
-	t.Skip("https://github.com/istio/istio/issues/22729")
-	rotateInterval := 1 * time.Second
 	sdsTest.RotateCert(rotateInterval)
 	setup := sdsTest.SetupTest(t, env.SDSCertRotation)
 	defer setup.TearDown()
-
 	setup.StartProxy(t)
 	start := time.Now()
-	numReq := 0
 	for {
 		code, _, err := env.HTTPGet(fmt.Sprintf("http://localhost:%d/echo", setup.OutboundListenerPort))
 		if err != nil {
 			t.Errorf("Failed in request: %v", err)
 		}
-		if code != 200 {
+		if code != httpSuccessCode {
 			t.Errorf("Unexpected status code: %d", code)
 		}
-		numReq++
-		if time.Since(start) > 4*rotateInterval {
+		if time.Since(start) > proxyRunningTime {
 			break
 		}
-		time.Sleep(100 * time.Millisecond)
-	}
-
-	stats, err := setup.ProxySetup.GetStatsMap()
-	if err == nil {
-		numSSLHandshake := stats["cluster.outbound_cluster_tls.ssl.handshake"]
-		numSSLConnError := stats[fmt.Sprintf("listener.127.0.0.1_%d.ssl.connection_error", setup.InboundListenerPort)]
-		numSSLVerifyNoCert := stats[fmt.Sprintf("listener.127.0.0.1_%d.ssl.fail_verify_no_cert", setup.InboundListenerPort)]
-		numSSLVerifyCAError := stats[fmt.Sprintf("listener.127.0.0.1_%d.ssl.fail_verify_error", setup.InboundListenerPort)]
-		numOutboundSDSUpdate := stats["cluster.outbound_cluster_tls.client_ssl_socket_factory.ssl_context_update_by_sds"]
-		numInboundSDSUpdate := stats[fmt.Sprintf("listener.127.0.0.1_%d.server_ssl_socket_factory.ssl_context_update_by_sds", setup.InboundListenerPort)]
-		// Cluster config max_requests_per_connection is set to 1, the number of requests should match
-		// the number of SSL connections. This guarantees SSL connection is using the latest TLS key/cert loaded in Envoy.
-		if numSSLHandshake != uint64(numReq) {
-			t.Errorf("Number of successful SSL handshake does not match, expect %d but get %d", numReq, numSSLHandshake)
-		}
-		if numSSLConnError != 0 {
-			t.Errorf("Number of SSL connection error: %d", numSSLConnError)
-		}
-		if numSSLVerifyNoCert != 0 {
-			t.Errorf("Number of SSL handshake failures because of missing client cert: %d", numSSLVerifyNoCert)
-		}
-		if numSSLVerifyCAError != 0 {
-			t.Errorf("Number of SSL handshake failures on CA verification: %d", numSSLVerifyCAError)
-		}
-		// Verify that there are multiple SDS updates. TLS key/cert are loaded multiple times.
-		if numOutboundSDSUpdate <= 1 {
-			t.Errorf("Number of SDS updates at outbound cluster should be greater than one, get %d", numOutboundSDSUpdate)
-		}
-		if numInboundSDSUpdate <= 1 {
-			t.Errorf("Number of SDS updates at inbound listener should be greater than one, get %d", numInboundSDSUpdate)
+		time.Sleep(sleepTime)
```

Comment:
This doesn't look safe. How do we know its long enough?

The response here should not be to make it sleep longer, since then the tests will be very slow. We should find a way to poll/signal readiness. Sleeping in unit tests is not safe, and will almost always result in flaky tests.

Reply:
Actually the total proxy running time will be about 4 times of the rotation time(which is 1 second), so for every 100ms , we will do a poll for the cert . So I think it is ok to set it here

Reply:
Add retry policy and long time interval for rotation